### PR TITLE
[FLINK-2026] add a flag to indicate previous executions

### PR DESCRIPTION
**Pull request by mxm**

### CODE REVIEW

**Comment by StephanEwen** [comment]
That makes sense. Can we rename the flag to `wasExecuted` or so?

How about a slight modification of the error message:

"No new data sinks have been defined since the last execution. The last execution refers to the latest call to 'execute()', 'count()' 'collect()' or 'print()'."

I added `print()` to that message as well, because it will become eager with a pending pull request.

**Comment by mxm** [comment]
- Yes, we can rename the variable to `wasExecuted`.
- I already added `print()` to the message because I anticipated it to become eager :)
- I like your error message, it's more concise.

**Comment by mxm** [comment]
Pull request updated.

File: flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java [hunk 1]
```diff
@@ -914,7 +917,15 @@ public JavaPlan createProgramPlan(String jobName) {
 	 */
 	public JavaPlan createProgramPlan(String jobName, boolean clearSinks) {
 		if (this.sinks.isEmpty()) {
-			throw new RuntimeException("No data sinks have been created yet. A program needs at least one sink that consumes data. Examples are writing the data set or printing it.");
+			if (wasExecuted) {
+				throw new RuntimeException("No new data sinks have been defined since the " +
+						"last execution. The last execution refers to the latest call to " +
+						"'execute()', 'count()' 'collect()' or 'print()'.");
```
Comment (uce) [inline flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java]: missing commas

**Comment by StephanEwen** [comment]
Looks good, modulo the commas.

+1 to merge after fixing this.

**Pull request closed without merging**